### ISSUE TITLE
facebookConnectPlugin.api returns the wrong results from other calls

### ISSUE THREAD

**Issue by digihaven** [Wizcorp/phonegap-facebook-plugin#1058]
If you call make calls like this:
''''''
	facebookConnectPlugin.api('/me', [], function(res) {
// Returns error
	},
	apiError);
	facebookConnectPlugin.api('/me/friends', ["user_friends"], function(res) {
// Returns results from /me!!!!
	},
	apiError);
''''
Doing this also has issues:
''''
	facebookConnectPlugin.api('/me/friends', ["user_friends"], function(res) {
// Returns error
	},
	apiError);
	facebookConnectPlugin.api('/me', [], function(res) {
// Returns results from /me/friends!!!!
	},
	apiError);
''''
Is there any reason why I get the wrong API call results?

**Comment by deejbee** [comment]
Did you discover what the problem with this is?  I think I may have a similar issue with paged results not returning correctly.  The first page is ok but when I try to get the next page its returns totally unexpected results.  Only on Android. OK on Web/Desktop.  Untested on iOS.

**Comment by digihaven** [comment]
Yes I did, and I rewrote the api() function. I could do a pull request but I have no idea if Wizcorp will take it since it's a massive rewrite.

**Comment by digihaven** [comment]
The quick fix for me is this (but it requires jQuery and should be removed for less bloat). It also fixed access tokens not working across sessions by adding them to localStorage. If there is enough interest I will fork and add to the main code with pull request.

````

				if (window.cordova.platformId === "browser")
				{
					facebookConnectPlugin.getAccessToken=function (s, f) {
						var response = FB.getAccessToken();
						response = response ? response : localStorage.getItem("FB.AccessToken");
						if (!response) {
							if (!f) {
								console.error("NO_TOKEN");
							} else {
								f("NO_TOKEN");
							}
						} else {
							localStorage.setItem("FB.AccessToken",response);
							s(response);
						}
					}

					facebookConnectPlugin.logout=function (s, f) {
					    // Try will catch errors when SDK has not been init
					    try {
						FB.logout( function (response) {
						    s(response);
						});
					    	localStorage.removeItem("FB.AccessToken");
					    } catch (error) {
						if (!f) {
						    console.error(error.message);
						} else {
						    f(error.message);
						}
					    }
					}
				}
                            }

				facebookConnectPlugin.api=function (graphPath, permissions, s, f) {
					// JS API does not take additional permissions
				    
					facebookConnectPlugin.getAccessToken(function(token) {
						console.log("Has access token: " + token);

						var q={access_token:token};
						$.ajax({
							url: "https://graph.facebook.com/v2.3" + graphPath +  (graphPath.match(/\?./) ? "&" : "?") + serialize(q)
						}).done(function(data) {
							s(data);
						}).error(function(data) {
							f(data.statusText);
						});
					}, function(err) {
						f(err);
					});
				}
````

**Comment by deejbee** [comment]
I did track down and did a quick fix (not committed- not sure how) with Android version in \platforms\android\src\org\apache\cordova\facebook\ConnectPlugin.java\makeGraphCall()

The iOS version also has the same problem and I don't know if this fixes all eventualities but feel free to use it.

     private void makeGraphCall() {
        Session session = Session.getActiveSession();

        Request.Callback graphCallback = new Request.Callback() {

            @Override
            public void onCompleted(Response response) {
                if (graphContext != null) {
                    if (response.getError() != null) {
                        graphContext.error(getFacebookRequestErrorResponse(response.getError()));
                    } else {
                        GraphObject graphObject = response.getGraphObject();
                        graphContext.success(graphObject.getInnerJSONObject());
                    }
                    graphPath = null;
                    graphContext = null;
                }
            }
        };

        //If you're using the paging URLs they will be URLEncoded, let's decode them.
        try {
            graphPath = URLDecoder.decode(graphPath, "UTF-8");
        } catch (UnsupportedEncodingException e) {
            e.printStackTrace();
        }

    //////////////////////////////////
    // start fix
    //////////////////////////////////
        String[] urlParts = graphPath.split("\\?");
        String graphAction = urlParts[0];
		
        final String GRAPH_BASE_URL = "https://graph.facebook.com/";
        if(graphAction.indexOf(GRAPH_BASE_URL)==0) {
            URL graphUrl = null;
            try {
                graphUrl = new URL(graphAction);
            } catch (MalformedURLException e) {
                e.printStackTrace();
            }
            graphAction = graphUrl.getPath();
        }


    //////////////////////////////////
    // end fix
    //////////////////////////////////

        Request graphRequest = Request.newGraphPathRequest(null, graphAction, graphCallback);
        Bundle params = graphRequest.getParameters();

        if (urlParts.length > 1) {
            String[] queries = urlParts[1].split("&");

            for (String query : queries) {
                int splitPoint = query.indexOf("=");
                if (splitPoint > 0) {
                    String key = query.substring(0, splitPoint);
                    String value = query.substring(splitPoint + 1, query.length());
                    params.putString(key, value);
                }
            }
        }
        params.putString("access_token", session.getAccessToken());

        graphRequest.setParameters(params);
        graphRequest.executeAsync();
    }
